tests/integration/test_upgrade.py

```diff
@@ -0,0 +1,126 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+#
+
+"""Integration tests for Seldon Core Operator/Charm Upgrade.
+
+Upgrade tests simulate in-field upgrade using local charms. A clean environment is required to test
+upgrade/refresh properly without cluttering other integraiton tests.
+"""
+
+import logging
+from pathlib import Path
+
+import pytest
+import yaml
+from lightkube import ApiError, Client
+from lightkube.resources.apiextensions_v1 import CustomResourceDefinition
+from lightkube.resources.core_v1 import ConfigMap
+from pytest_operator.plugin import OpsTest
+
+logger = logging.getLogger(__name__)
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+APP_NAME = "seldon-controller-manager"
+
+
+# Skip upgrade test, because it is failing in CI due to authorization issues.
+# Manual test instructions for upgrade are provided in corresponding Github issue:
+# https://github.com/canonical/seldon-core-operator/issues/101
+# Upgrade test can be executed locally.
+# TO-DO Ensure upgrade test passes in CI environment.
+@pytest.mark.skip(reason="Skip due to authorization issues in CI.")
+@pytest.mark.abort_on_fail
+async def test_upgrade(ops_test: OpsTest):
+    """Test upgrade.
+
+    Verify that all upgrade process succeeds.
+
+    NOTE: There should be no charm with APP_NAME deployed prior to testing upgrade, because this
+    test deploys local stable version of this charm which has revision 0 and peforms upgrade to
+    locally built charm which will have revision 1. If prior deployment of locally build or remote
+    charm is done, initial revision will be different and this mismatch will cause upgrade to fail.
+    There should be no Seldon related resources present in the cluster.
+
+    Main flow of the test:
+    - Build charm to be tested, i.e. to be upgraded to.
+    - Download and deploy stable version of the same charm and store it in the same location as the
+      charm to be tested (juju refresh of local charm requires local charms to be in the same
+      path). Note that stable/1.14 version should be deployed with "trust"
+    - Refresh the deployed stable charm to the charm to be tested.
+    - Verify that charm is active and all resources are upgraded/intact.
+    """
+
+    # build the charm
+    charm_under_test = await ops_test.build_charm(".")
+
+    # download and deploy stable version of the charm
```

Comment:
Why do you need to download the charm? is `ops_test.model.deploy()` not working for you?

Reply:
I am downloading stable charm from Charmhub. Then deploying it with `.deploy()`, then upgrading it with to locally built with `.refresh()`

Reply:
I understand what you are doing, but I don't understand why you download it instead of just deploying it.

Reply:
I am downloading stable charm (`--channel=1.14/stable`) from 1.6, so I can upgrade it to the current, most latest one we are building/testing in this test.

Reply:
Waiting on comments from Juju team, will update according to their suggested.